[kernel 3a/n] Decouple `CTxMemPool` from `ArgsManager`

dongcarl opened the pull request:
This is part of the `libbitcoinkernel` project: #24303, https://github.com/bitcoin/bitcoin/projects/18

-----

As mentioned in the Stage 1 Step 2 description of [the `libbitcoinkernel` project](https://github.com/bitcoin/bitcoin/issues/24303), `ArgsManager` will not be part of `libbitcoinkernel`. Therefore, it is important that we remove any dependence on `ArgsManager` by code that will be part of `libbitcoinkernel`. This is the first in a series of PRs aiming to achieve this.

This PR removes `CTxMemPool+MempoolAccept`'s dependency on `ArgsManager` by introducing a `CTxMemPool::Options` struct, which is used to specify `CTxMemPool`'s various options at construction time.

These options are:
- `-maxmempool` -> `CTxMemPool::Options::max_size`
- `-mempoolexpiry` -> `CTxMemPool::Options::expiry`
- `-limitancestorcount` -> `CTxMemPool::Options::limits::ancestor_count`
- `-limitancestorsize` -> `CTxMemPool::Options::limits::ancestor_size`
- `-limitdescendantcount` -> `CTxMemPool::Options::limits::descendant_count`
- `-limitdescendantsize` -> `CTxMemPool::Options::limits::descendant_size`

More context can be gleaned from the commit messages. The important commits are:

- 56eb479ded8bfb2ef635bb6f3b484f9d5952c70d "pool: Add and use MemPoolOptions, ApplyArgsManOptions"
- a1e08b70f3068f4e8def1c630d8f50cd54da7832 "mempool: Pass in -maxmempool instead of referencing gArgs"
- 6f4bf3ede5812b374828f08fc728ceded2f10024 "mempool: Pass in -mempoolexpiry instead of referencing gArgs"
- 5958a7fe4806599fc620ee8c1a881ca10fa2dd16 "mempool: Introduce (still-unused) MemPoolLimits"

Reviewers: Help needed in the following commits (see commit messages):
- a1e08b70f3068f4e8def1c630d8f50cd54da7832 "mempool: Pass in -maxmempool instead of referencing gArgs"
- 0695081a797e9a5d7787b78b0f8289dafcc6bff7 "node/ifaces: Use existing MemPoolLimits"

Note to Reviewers: There are perhaps an infinite number of ways to architect `CTxMemPool::Options`, the current one tries to keep it simple, usable, and flexible. I hope we don't spend too much time arguing over the design here since that's not the point. In the case that you're 100% certain that a different design is strictly better than this one in every regard, please show us a fully-implemented branch.

-----

TODO:
- [x] Use the more ergonomic `CTxMemPool::Options` where appropriate
- [x] Doxygen comments for `ApplyArgsManOptions`, `MemPoolOptions`

-----

Questions for Reviewers:
1. Should we use `std::chrono::seconds` for `CTxMemPool::Options::expiry` and `CTxMemPool::m_expiry` instead of an `int64_t`? Something else? (`std::chrono::hours`?)
2. Should I merge `CTxMemPool::Limits` inside `CTxMemPool::Options`?

dongcarl committed:
ArgsMan: Add Get*Arg functions returning optional

This allows the caller to not provide a default at all and just check
inside the optional to see if the arg was set or not.

dongcarl committed:
mempool: Pass in -maxmempool instead of referencing gArgs

- Store the mempool size limit (-maxmempool) in CTxMemPool as a member.

- Remove the requirement to explicitly specify a mempool size limit for
  CTxMemPool::GetMinFee(...) and LimitMempoolSize(...), just use the
  stored mempool size limit where possible.

- Remove all now-unnecessary instances of:
    gArgs.GetIntArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE_MB) * 1000000

The code change in CChainState::GetCoinsCacheSizeState() is correct
since the coinscache should not repurpose "extra" mempool memory
headroom for itself if the mempool doesn't even exist.

dongcarl committed:
mempool: Pass in -mempoolexpiry instead of referencing gArgs

- Store the mempool expiry (-mempoolexpiry) in CTxMemPool as a
  std::chrono::seconds member.

- Remove the requirement to explicitly specify a mempool expiry for
  LimitMempoolSize(...), just use the newly-introduced member.

- Remove all now-unnecessary instances of:
    std::chrono::hours{gArgs.GetIntArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY)}

dongcarl committed:
fees: Pass in a filepath instead of referencing gArgs

dongcarl committed:
scripted-diff: Rename anc/desc size limit vars to indicate SI unit

Better to be explicit when it comes to sizes to avoid unintentional
bugs. We use MB and KB all over the place.

-BEGIN VERIFY SCRIPT-
find_regex="DEFAULT_(ANCESTOR|DESCENDANT)_SIZE_LIMIT" \
    && git grep -l -E "$find_regex" \
        | xargs sed -i -E "s@$find_regex@\0_KVB@g"
-END VERIFY SCRIPT-

dongcarl committed:
mempool: Introduce (still-unused) MemPoolLimits

They live as a CTxMemPool member.

[META] These limits will be used in subsequent commits to replace calls
       to gArgs.

dongcarl committed:
pool: Add and use MemPoolOptions, ApplyArgsManOptions

Reviewers: Note that CTxMemPool now requires a non-defaulted
CTxMemPool::Options for its constructor. Meaning that there's no need to
worry about a stray CTxMemPool constructor somewhere defaulting to
something incorrect. All instances of CTxMemPool construction are
addressed here in this commit.

We set options for CTxMemPool and construct it in many different ways. A
good example can be seen in how we determine CTxMemPool's check_ratio in
AppInitMain(...).

1. We first set the default based on chainparams's
   DefaultConsistencyChecks()
2. Then, we apply the ArgsManager option on top of that default
3. Finally, we clamp the result of that between 0 and 1 Million

With this patch, most CTxMemPool construction are along the lines of:

    MemPoolOptions mempool_opts{...default overrides...};
    ApplyArgsManOptions(argsman, mempool_opts);
    ...hard overrides...
    CTxMemPool pool{mempool_opts};

This "compositional" style of building options means that we can omit
unnecessary/irrelevant steps wherever we want but also maintain full
customizability.

For example:

- For users of libbitcoinkernel, where we eventually want to remove
  ArgsManager, they simply won't call (or even know about)
  ApplyArgsManOptions.

- See src/init.cpp to see how the check_ratio CTxMemPool option works
  after this change.

A MemPoolOptionsForTest helper was also added and used by tests/fuzz
tests where a local CTxMemPool needed to be created.

The change in src/test/fuzz/tx_pool.cpp seemingly changes behaviour by
applying ArgsManager options on top of the CTxMemPool::Options defaults.
However, in future commits where we introduce flags like -maxmempool,
the call to ApplyArgsManOptions is actually what preserves the existing
behaviour. Previously, although it wasn't obvious, our CTxMemPool would
consult gArgs for flags like -maxmempool when it needed it, so it
already relied on ArgsManager information. This patchset just laid bare
the obfuscatory perils of globals.

[META] As this patchset progresses, we will move more and more
       CTxMemPool-relevant options into MemPoolOptions and add their
       ArgsMan-related logic to ApplyArgsManOptions.

dongcarl committed:
init: Only determine maxmempool once

Now that MemPoolOptions has a correctly-determined max_size member, use
that instead of redetermining it to print the log line.

dongcarl committed:
scripted-diff: Rename DEFAULT_MAX_MEMPOOL_SIZE to indicate SI unit

Better to be explicit when it comes to sizes to avoid unintentional
bugs. We use MB and KB all over the place.

-BEGIN VERIFY SCRIPT-
find_regex="DEFAULT_MAX_MEMPOOL_SIZE" \
    && git grep -l -E "$find_regex" \
        | xargs sed -i -E "s@$find_regex@\0_MB@g"
-END VERIFY SCRIPT-

dongcarl committed:
node/ifaces: Use existing MemPoolLimits

dongcarl committed:
mempool: Use m_limit for UpdateTransactionsFromBlock

Since:

- UpdateTransactionsFromBlock is only called by
  MaybeUpdateMempoolForReorg, which calls it with the gArgs-determined
  ancestor limits
- UpdateForDescendants is only called by UpdateTransactionsFromBlock
  with the ancestor limits unchanged

We can remove the requirement to specify the ancestor limits for both
UpdateTransactionsFromBlock and UpdateForDescendants and just use the
values in the m_limits member.

Also move some removed comments to MemPoolLimits struct members.

The uint64_t cast in UpdateForDescendants is not new behavior,
see the diff in CChainState::MaybeUpdateMempoolForReorg for where they
were previously.

dongcarl committed:
init: Remove redundant -*mempool*, -limit* queries

Now that MemPoolOptions has correctly-determined max_size and limits
members, perform sanity checks on that instead of re-determining the
options.

dongcarl committed:
mempoolaccept: Use limits from mempool in constructor

dongcarl commented:
Pushed 532f83872d3921030fc4240fa8a8c4f6a1aaacf0 -> 4bbc2e0cfc2e395649a20e616648937c81dca377
- Added doxygen comments for `MemPoolOptions`, `MemPoolLimits`, `ApplyArgsManOptions`
- Use `CTxMemPool::{Options,Limits}` instead of `MemPool{Options,Limits}` wherever possible
- Move `ApplyArgsManOptions(const ArgsManager&, MemPoolLimits&)` to an anonymous namespace in  `mempool_args.cpp`

dongcarl commented:
Pushed f803134e4cf1f2566858e0c597fcd19e29afed3f -> c98d5f1744dc3beed116ff708e3d4d8dcfe53eae
- Rebased after merge of #25215 and #25223 (thanks all reviewers!)
- Marking as Ready for Review as all dependent PRs have been merged

fanquake commented:
https://cirrus-ci.com/task/6709025164230656:
```bash
Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_node.py", line 552, in assert_start_raises_init_error
                                       ret = self.process.wait(timeout=self.rpc_timeout)
                                     File "/usr/lib/python3.6/subprocess.py", line 1469, in wait
                                       raise TimeoutExpired(self.args, timeout)
                                   subprocess.TimeoutExpired: Command '['/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/src/bitcoind', '-datadir=/tmp/cirrus-ci-build/ci/scratch/test_runner/test_runner_\u20bf_🏃_20220616_191835/mempool_limit_214/node0', '-logtimemicros', '-debug', '-debugexclude=libevent', '-debugexclude=leveldb', '-uacomment=testnode0', '-logthreadnames', '-logsourcelocations', '-maxmempool=4']' timed out after 2400 seconds
                                   During handling of the above exception, another exception occurred:
                                   Traceback (most recent call last):
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 133, in main
                                       self.run_test()
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/mempool_limit.py", line 82, in run_test
                                       self.nodes[0].assert_start_raises_init_error(["-maxmempool=4"], "Error: -maxmempool must be at least 5 MB")
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_node.py", line 582, in assert_start_raises_init_error
                                       self._raise_assertion_error(assert_msg)
                                     File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_node.py", line 167, in _raise_assertion_error
                                       raise AssertionError(self._node_msg(msg))
                                   AssertionError: [node 0] bitcoind should have exited within 2400s with expected error Error: -maxmempool must be at least 5 MB
```

glozow commented on src/node/interfaces.cpp:
```diff
@@ -653,24 +654,25 @@ class ChainImpl : public Chain
     {
-        limit_ancestor_count = gArgs.GetIntArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT);
-        limit_descendant_count = gArgs.GetIntArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT);
+        limit_ancestor_count = limit_descendant_count = 0;
+        if (!m_node.mempool) return;
```
in 95430b611c6562845143669e88e07fcde770d1af "node/ifaces: Use existing MemPoolLimits"

I see your "Question for Reviewers: Is the default for limit_ancestor_count and limit_descendant_count in getPackageLimits correct?" in the commit message. To clarify, are you asking whether this has a behavior change, or are you asking if the change is better?

IIUC This returns limits = 0 when there is no mempool, where it would have previously returned the results of querying gArgs / defaults. This makes logical sense to me: if there's no mempool, then there's no mempool limits. It seems the one use for `getPackageLimits` is in `SelectCoins`, and the wallet understands it can still make an unconfirmed transaction (i.e. anc/desc count = 1) so it's fine: https://github.com/bitcoin/bitcoin/blob/874529665c1c326fc86fc0d0d6c3512fab087ab8/src/wallet/spend.cpp#L509-L511

I guess a future `-nomempool` option would just need to be consistent with this.

dongcarl replied:
Fixed in 4825207fc6c15fce4cc1cb9b14a66a51f3b5a47b, feel free to reopen if still a problem!

glozow commented on src/init.cpp:
```diff
@@ -932,10 +932,6 @@ bool AppInitParameterInteraction(const ArgsManager& args, bool use_syscall_sandb
 
     // mempool limits
-    int64_t nMempoolSizeMax = args.GetIntArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE_MB) * 1000000;
-    int64_t nMempoolSizeMin = args.GetIntArg("-limitdescendantsize", DEFAULT_DESCENDANT_SIZE_LIMIT_KB) * 1000 * 40;
```
In efcbd12767d87b33b6c6deb86d441292ff681b13 "init: Remove redundant -*mempool*, -limit* queries "

Seems like you've dropped a `* 40` here? Is that maybe what's causing the CI failure?

glozow reviewed (commented):
Concept ACK

Re: "Should I merge CTxMemPool::Limits inside CTxMemPool::Options?"

(TLDR imo keeping them separate is fine).
Would it be accurate to think of `Limits` as policy (used externally in `MemPoolAccept`) while `Options` is related to managing the data structure of `CTxMemPool` itself (i.e. it knows its own check ratio and max size)? Imo separating the ancestor/descendant limits makes sense because they're not just internal to `CTxMemPool`; they're also used by `MemPoolAccept` to implement carveouts. For example, it will override them by adding +1 to the descendant limit and setting ancestor limit = 2 for CPFP carve out: https://github.com/bitcoin/bitcoin/blob/874529665c1c326fc86fc0d0d6c3512fab087ab8/src/validation.cpp#L906-L908

ryanofsky reviewed (approved):
Conditional code review ACK c98d5f1744dc3beed116ff708e3d4d8dcfe53eae if CI error is fixed. Feel free to ignore all the suggestions, this basically looks good.

darosior commented on src/mempool_options.h:
```diff
@@ -4,8 +4,13 @@
 class CBlockPolicyEstimator;
 
+/** Default for -maxmempool, maximum megabytes of mempool memory usage */
+static const unsigned int DEFAULT_MAX_MEMPOOL_SIZE_MB = 300;
```
nit: maybe make it (along with the expiry one) `constexpr`?

dongcarl replied:
Fixed in 52349ea72f58b4057a88890fe0c7b722698e3fac

darosior reviewed (commented):
Concept ACK. I'd like to take another look at the mempool limits changes, but i agree it looks good modulo the dropped `* 40` pointed by Gloria.

> Should I merge `CTxMemPool::Limits` inside `CTxMemPool::Options`?

Rebasing after #25388 makes introducing `mempool_limits.h` just for the `MemPoolLimits` struct, but on the other hand those limits really are policy and i agree with Gloria it's preferable to not mix them with options internal to the mempool. So i too think it's fine as is.

ariard reviewed (commented):
SGTM modulo I think this [comment](https://github.com/bitcoin/bitcoin/pull/25290#discussion_r900059562) and this other [one](https://github.com/bitcoin/bitcoin/pull/25290#discussion_r900405188) to fix.

> Should I merge CTxMemPool::Limits inside CTxMemPool::Options?

I think it's better to keep them split as it might be interesting to give the transaction/package-related limits to the wallet component, even in the lack of a mempool. It would solve the discussion around `getPackageLimits()`.

dongcarl commented on src/init.cpp:
```diff
@@ -539,9 +539,9 @@ void SetupServerArgs(ArgsManager& argsman)
     argsman.AddArg("-stopatheight", strprintf("Stop running after reaching the given height in the main chain (default: %u)", DEFAULT_STOPATHEIGHT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
     argsman.AddArg("-limitancestorcount=<n>", strprintf("Do not accept transactions if number of in-mempool ancestors is <n> or more (default: %u)", DEFAULT_ANCESTOR_LIMIT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
-    argsman.AddArg("-limitancestorsize=<n>", strprintf("Do not accept transactions whose size with all in-mempool ancestors exceeds <n> kilobytes (default: %u)", DEFAULT_ANCESTOR_SIZE_LIMIT), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
+    argsman.AddArg("-limitancestorsize=<n>", strprintf("Do not accept transactions whose size with all in-mempool ancestors exceeds <n> kilobytes (default: %u)", DEFAULT_ANCESTOR_SIZE_LIMIT_KB), ArgsManager::ALLOW_ANY | ArgsManager::DEBUG_ONLY, OptionsCategory::DEBUG_TEST);
```
Fixed in 67ad3f84b1e69ad830ee27b25e0b6a33519e1d0b

dongcarl commented on src/mempool_limits.h:
```diff
@@ -0,0 +1,31 @@
+ */
+struct MemPoolLimits {
+    size_t ancestor_count = DEFAULT_ANCESTOR_LIMIT;
+    size_t ancestor_size = DEFAULT_ANCESTOR_SIZE_LIMIT_KB * 1000;
```
Fixed in 291b605e21b5f4eb739cd20418ca1fc6f467fcab

dongcarl commented on src/test/util/setup_common.cpp:
```diff
@@ -163,7 +163,7 @@ ChainTestingSetup::ChainTestingSetup(const std::string& chainName, const std::ve
     GetMainSignals().RegisterBackgroundSignalScheduler(*m_node.scheduler);
 
-    m_node.fee_estimator = std::make_unique<CBlockPolicyEstimator>();
+    m_node.fee_estimator = std::make_unique<CBlockPolicyEstimator>(gArgs.GetDataDirNet() / FEE_ESTIMATES_FILENAME);
```
Fixed in 125c198efe1b2157a669758e7dadd1d15fbec368

dongcarl commented on src/txmempool.h:
```diff
@@ -560,15 +561,14 @@ class CTxMemPool
      * in the pool.
-     *
-     * @param[in] estimator is used to estimate appropriate transaction fees.
-     * @param[in] check_ratio is the ratio used to determine how often sanity checks will run.
```
Fixed in 7e6d2d72368631b03d00c8fc7689a670ba92e4ac

dongcarl commented on src/validation.h:
```diff
@@ -61,20 +61,20 @@ struct Params;
 static const unsigned int DEFAULT_DESCENDANT_LIMIT = 25;
 /** Default for -limitdescendantsize, maximum kilobytes of in-mempool descendants */
-static const unsigned int DEFAULT_DESCENDANT_SIZE_LIMIT = 101;
+static const unsigned int DEFAULT_DESCENDANT_SIZE_LIMIT_KB = 101;
```
Fixed in 67ad3f84b1e69ad830ee27b25e0b6a33519e1d0b

dongcarl commented on src/mempool_options.h:
```diff
@@ -22,6 +25,7 @@ struct MemPoolOptions {
     CBlockPolicyEstimator* const estimator = nullptr;
     int check_ratio = 0;
     size_t max_size = DEFAULT_MAX_MEMPOOL_SIZE_MB * 1000000;
+    int64_t expiry = DEFAULT_MEMPOOL_EXPIRY;
```
How about `std::chrono::seconds`? It seems that's how it's used/compared in our code anyway

MarcoFalke replied:
It could be `NodeClock::duration` with the reason that the value is compared against the value of `GetTime()`, which is deprecated in favour of (and an alias to) `NodeClock::now()`.

dongcarl committed:
scripted-diff: Rename DEFAULT_MEMPOOL_EXPIRY to indicate time unit

Better to be explicit when it comes to time to avoid unintentional bugs.

-BEGIN VERIFY SCRIPT-
find_regex="DEFAULT_MEMPOOL_EXPIRY" \
    && git grep -l -E "$find_regex" \
        | xargs sed -i -E "s@$find_regex@\0_HOURS@g"
-END VERIFY SCRIPT-

ariard reviewed (approved):
Code Review ACK 125c198e

Changes since last review:
- add MB suffix to `DEFAULT_MAX_MEMPOOL_SIZE`
- add comments in `MemPoolOptions`
- one-lining `mempool_max_size_mb` assignment in `ApplyArgsManOptions`
- add bytes suffix to `max_size`
- make `DEFAULT_MEMPOOL_EXPIRY` constexpr
- add vbytes suffix to `ancestor/descendant_size`
- change SI of `DEFAULT_ANCESTOR_SIZE_LIMIT` from KB to KVB
- return default package limits in lack of mempool in `getPackageLimits()`
- use correct `descendant_limit_bytes` to evaluate min mempool max size

I don't know about the compilation warning, if you get ride of it in another `libbitcoinkernel` PR.

dongcarl commented on src/mempool_options.h:
```diff
@@ -16,6 +21,7 @@ class CBlockPolicyEstimator;
 struct MemPoolOptions {
     CBlockPolicyEstimator* const estimator = nullptr;
     int check_ratio = 0;
+    size_t max_size = DEFAULT_MAX_MEMPOOL_SIZE_MB * 1000000;
```
Fixed in 8c81736608e05607d0d84a6edb7ef6b66e46857c

dongcarl commented on src/node/interfaces.cpp:
```diff
@@ -653,24 +654,25 @@ class ChainImpl : public Chain
+        const CTxMemPool::Limits& limits{m_node.mempool->m_limits};
         std::string unused_error_string;
         LOCK(m_node.mempool->cs);
         return m_node.mempool->CalculateMemPoolAncestors(
```
Do you mean something like below? This convenience method would really only benefit this one call site due to the limitations I mentioned here: https://github.com/bitcoin/bitcoin/pull/25290#discussion_r902840978

```diff
diff --git a/src/node/interfaces.cpp b/src/node/interfaces.cpp
index cd2ad1f6a2..9977968775 100644
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -674,8 +674,7 @@ public:
         std::string unused_error_string;
         LOCK(m_node.mempool->cs);
         return m_node.mempool->CalculateMemPoolAncestors(
-            entry, ancestors, limits.ancestor_count, limits.ancestor_size_vbytes,
-            limits.descendant_count, limits.descendant_size_vbytes, unused_error_string);
+            entry, ancestors, limits, unused_error_string);
     }
     CFeeRate estimateSmartFee(int num_blocks, bool conservative, FeeCalculation* calc) override
     {
diff --git a/src/txmempool.cpp b/src/txmempool.cpp
index baf5718a8d..878e828b65 100644
--- a/src/txmempool.cpp
+++ b/src/txmempool.cpp
@@ -295,6 +295,20 @@ bool CTxMemPool::CheckPackageLimits(const Package& package,
     return ret;
 }
 
+bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry& entry,
+                                           setEntries& setAncestors,
+                                           const CTxMemPool::Limits& limits,
+                                           std::string& errString,
+                                           bool fSearchForParents /* = true */) const
+{
+    return CalculateMemPoolAncestors(entry, setAncestors,
+                                     limits.ancestor_count,
+                                     limits.ancestor_size_vbytes,
+                                     limits.descendant_count,
+                                     limits.descendant_size_vbytes,
+                                     errString, fSearchForParents);
+}
+
 bool CTxMemPool::CalculateMemPoolAncestors(const CTxMemPoolEntry &entry,
                                            setEntries &setAncestors,
                                            uint64_t limitAncestorCount,
diff --git a/src/txmempool.h b/src/txmempool.h
index 76f51e9d1e..a71c80e2b8 100644
--- a/src/txmempool.h
+++ b/src/txmempool.h
@@ -669,7 +669,19 @@ public:
      *  fSearchForParents = whether to search a tx's vin for in-mempool parents, or
      *    look up parents from mapLinks. Must be true for entries not in the mempool
      */
-    bool CalculateMemPoolAncestors(const CTxMemPoolEntry& entry, setEntries& setAncestors, uint64_t limitAncestorCount, uint64_t limitAncestorSize, uint64_t limitDescendantCount, uint64_t limitDescendantSize, std::string& errString, bool fSearchForParents = true) const EXCLUSIVE_LOCKS_REQUIRED(cs);
+    bool CalculateMemPoolAncestors(const CTxMemPoolEntry& entry,
+                                   setEntries& setAncestors,
+                                   const CTxMemPool::Limits& limits,
+                                   std::string& errString,
+                                   bool fSearchForParents = true) const EXCLUSIVE_LOCKS_REQUIRED(cs);
+    bool CalculateMemPoolAncestors(const CTxMemPoolEntry& entry,
+                                   setEntries& setAncestors,
+                                   uint64_t limitAncestorCount,
+                                   uint64_t limitAncestorSize,
+                                   uint64_t limitDescendantCount,
+                                   uint64_t limitDescendantSize,
+                                   std::string& errString,
+                                   bool fSearchForParents = true) const EXCLUSIVE_LOCKS_REQUIRED(cs);
 
     /** Calculate all in-mempool ancestors of a set of transactions not already in the mempool and
      * check ancestor and descendant limits. Heuristics are used to estimate the ancestor and
```

dongcarl replied:
Lots of call sites so not going to do in this PR, added to project-wide TODOs!

fanquake replied:
The move to using a `Limits` struct for `CalculateMemPoolAncestors` was done in #26103.

dongcarl commented:
Push c98d5f1744dc3beed116ff708e3d4d8dcfe53eae -> 167076739f02ddcf754200b378d504d320d0ff22
- Resolved most of the review comments (Thanks Gloria, Ryan, and Antoine^2) ☺️

dongcarl commented on src/mempool_limits.h:
```diff
@@ -0,0 +1,23 @@
+ * Most of the time, this struct should be referenced as CTxMemPool::Limits.
+ */
+struct MemPoolLimits {
+    size_t ancestor_count = DEFAULT_ANCESTOR_LIMIT;
+    size_t ancestor_size_vbytes = DEFAULT_ANCESTOR_SIZE_LIMIT_KVB * 1000;
+    size_t descendant_count = DEFAULT_DESCENDANT_LIMIT;
+    size_t descendant_size_vbytes = DEFAULT_DESCENDANT_SIZE_LIMIT_KVB * 1000;
```
Reviewers: What do people think about using `uint64_t` here instead of `size_t`? It seems like there's no obvious need for these to be platform-specific and it might be better to not have to worry about them being inconsistent across platforms.

laanwj replied:
You have a point, but I don't know. All the array/vector size functions return size_t, right? So making these `uint64_t` introduces an implicit cast. That while they're unlikely to ever be outside the range of `uint32_t` anyway.

glozow replied:
> What do people think about using uint64_t here instead of size_t? It seems like there's no obvious need for these to differ from platform to platform and it might be better to not have to worry about them being inconsistent across platforms.

Yeah makes sense. Do you mean to do those in this PR? It seems there are thoughts about changing all virtual size representation to int64_t or int32_t across the board (signed because we do arithmetic with them). See #23962 and #23957

dongcarl replied:
See: https://github.com/bitcoin/bitcoin/pull/25290#issuecomment-1165951176

dongcarl replied:
Made these `int64_t`s to avoid an unintended change in behavior. Added a project-wide TODO to have bounds checking in `ApplyArgsManOptions`

MarcoFalke commented on src/Makefile.am:
```diff
@@ -178,6 +178,9 @@ BITCOIN_CORE_H = \
   mapport.h \
+  mempool_args.h \
+  mempool_limits.h \
+  mempool_options.h \
```
chainstatemanager_opts.h is located in `kernel/`, but this not. why?

MarcoFalke commented on src/mempool_args.cpp:
```diff
@@ -0,0 +1,32 @@
+namespace {
+void ApplyArgsManOptions(const ArgsManager& argsman, MemPoolLimits& mempool_limits)
+{
+    mempool_limits.ancestor_count = argsman.GetIntArg("-limitancestorcount", mempool_limits.ancestor_count);
```
nit in 6843403a15cab34bc4e1e7acf7a0e509a2d437df:

```suggestion
    if (auto cnt = argsman.GetIntArg("-limitancestorcount") mempool_limits.ancestor_count = *cnt;
```

Any reason to use a different style for those, when it is less code to keep the style consistent with the other parsing?

dongcarl replied:
I wanted to make visually clear the cases where we directly use the `ArgsMan` arg in `CTxMemPool::Options` and which cases get transformed before getting set in `CTxMemPool::Options`. However, I don't feel strongly so let me know if you prefer the consistency instead!

MarcoFalke replied:
I don't care too much. My suggestion uses less code, but this is only a style nit.

MarcoFalke commented on src/mempool_args.cpp:
```diff
@@ -0,0 +1,32 @@
+
+void ApplyArgsManOptions(const ArgsManager& argsman, MemPoolOptions& mempool_opts)
+{
+    mempool_opts.check_ratio = argsman.GetIntArg("-checkmempool", mempool_opts.check_ratio);
```
same nit:

```suggestion
    if (auto ratio = argsman.GetIntArg("-checkmempool") mempool_opts.check_ratio = *ratio;
```

dongcarl replied:
See https://github.com/bitcoin/bitcoin/pull/25290#discussion_r905398043

MarcoFalke commented on src/node/interfaces.cpp:
```diff
@@ -653,24 +654,28 @@ class ChainImpl : public Chain
     {
-        limit_ancestor_count = gArgs.GetIntArg("-limitancestorcount", DEFAULT_ANCESTOR_LIMIT);
-        limit_descendant_count = gArgs.GetIntArg("-limitdescendantcount", DEFAULT_DESCENDANT_LIMIT);
+        if (!m_node.mempool) {
+            const CTxMemPool::Limits limits{};
+            limit_ancestor_count = limits.ancestor_count;
+            limit_descendant_count = limits.descendant_count;
+        } else {
+            const CTxMemPool::Limits& limits{m_node.mempool->m_limits};
+            limit_ancestor_count = limits.ancestor_count;
+            limit_descendant_count = limits.descendant_count;
+        }
```
nit in b810f102c47c61f276ffa2f9a78a3b86ad5a88b8 for less code:

```suggestion
            const CTxMemPool::Limits default_limits{};
            
            const CTxMemPool::Limits& limits{m_node.mempool?m_node.mempool->m_limits:default_limits};
            limit_ancestor_count = limits.ancestor_count;
            limit_descendant_count = limits.descendant_count;
```

dongcarl replied:
Fixed as of d669b93469

MarcoFalke commented on src/init.cpp:
```diff
@@ -932,10 +932,6 @@ bool AppInitParameterInteraction(const ArgsManager& args, bool use_syscall_sandb
     }
 
     // mempool limits
```
7e90c69f1be6f8ba52e1f24b983379d00aad2f65 : forgot to remove this?

dongcarl replied:
Fixed as of a956afedfe

MarcoFalke commented on src/init.cpp:
```diff
@@ -1434,6 +1430,11 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     mempool_opts.check_ratio = std::clamp<int>(mempool_opts.check_ratio, 0, 1000000);
+
+    size_t descendant_limit_bytes = mempool_opts.limits.descendant_size_vbytes * 40;
+    if (mempool_opts.max_size_bytes < 0 || mempool_opts.max_size_bytes < descendant_limit_bytes) {
```
7e90c69f1be6f8ba52e1f24b983379d00aad2f65:

this doesn't work. You can't parse a signed integer and assign it to an unsigned type, then use `< 0` to check if it originally was negative.

ariard replied:
Already reported here : https://github.com/bitcoin/bitcoin/pull/25290#discussion_r902827767, get caught as a compilation warning.

MarcoFalke replied:
To solve, it would either need to be cast back with `int64_t(...) < 0` or the type changed.

MarcoFalke commented on src/init.cpp:
```diff
@@ -1434,6 +1430,11 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+
+    size_t descendant_limit_bytes = mempool_opts.limits.descendant_size_vbytes * 40;
+    if (mempool_opts.max_size_bytes < 0 || mempool_opts.max_size_bytes < descendant_limit_bytes) {
+        return InitError(strprintf(_("-maxmempool must be at least %d MB"), std::ceil(descendant_limit_bytes / 1000000.0)));
```
7e90c69f1be6f8ba52e1f24b983379d00aad2f65: Maybe use `1'000'000.0` (not sure if that compiles?)

dongcarl replied:
It does! Fixed as of a956afedfe

MarcoFalke commented on src/policy/fees.h:
```diff
@@ -19,6 +20,8 @@
 #include <string>
 #include <vector>
 
+extern const char* FEE_ESTIMATES_FILENAME;
```
167076739f02ddcf754200b378d504d320d0ff22:

Instead of exporting this, what about exporting a function that calculates the path?

```
fs::path FeeestPath(ArgsManager&);
```

So caller could call `FeeestPath(gArgs)` instead of `gArgs.GetDataDirNet() / FEE_ESTIMATES_FILENAME`

MarcoFalke replied:
Closing as resolved

maflcko reviewed (approved):
Looks very good, but I think commit 7e90c69f1be6f8ba52e1f24b983379d00aad2f65 doesn't work.

review ACK 167076739f02ddcf754200b378d504d320d0ff22 📽

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 167076739f02ddcf754200b378d504d320d0ff22 📽
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh5rAwApjNyIo4vWTlBPYX0A7LPHtMuhGjhptrfleuzrHmFNmrtt0af0IF4wdDw
Q2p4O4h2b+tdGfwahr4aW9RJ8/UXUkyNnLDtRhiGwl3Km5u0wWhgMtBbCAoIDs9k
oBbOnSlG1Fry+sey7dPCpvFcfcPBV3ldPvG6M1Bi2lWWFrLVVCnIsQYzUfC52Roz
RIsBtEr2u7kX0i3cmklWpSps25yDSIWnwxCbrwnqfVYfw2ntZN7PGT3PdIu6kFuK
qcAwEHe7GR3hZjCVlr2uo9QPGbTNFYhPX+zgjDjZVATmoR0rqL++bVrv/EiGBcNZ
QApyOuattxjPBbqBYgj8WQ2rYr/QfB+1WPMCmmibtC5ghkfFRiu5iEvrKqicU7f4
ZpRdH/chVY4xUvXbDgGBevaCDW6m0Jlu2Y0TVYXA7HzDP1novRtr7qQoUBhJW+f6
WAmRfwf02VjVRGLbeTesbTpKnZ0/u6R4RZI41xlaGtB/KgMt3S72tvbby/De9aEI
YkeDfTK9
=P9lh
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented on src/mempool_options.h:
```diff
@@ -0,0 +1,37 @@
+ */
+struct MemPoolOptions {
+    /* Used to estimate appropriate transaction fees. */
+    CBlockPolicyEstimator* const estimator = nullptr;
```
For all commits: Any reason to not use `{}` over `=` in this struct for C++11 member inits?

dongcarl replied:
Nope! Is `{}` preferred?

MarcoFalke replied:
Yes, for new code it is preferred style. `{}` has narrowing warnings enabled, as opposed to `=` or `()`, which do not.

MarcoFalke replied:
Closing as resolved

w0xlt reviewed (approved):
Code Review ACK https://github.com/bitcoin/bitcoin/pull/25290/commits/167076739f02ddcf754200b378d504d320d0ff22

. `MemPoolOptions`, `MemPoolLimits` and `ApplyArgsManOptions` make the mempool code more modularized.
. Adding SI and time units improves code readability.
. Replacing `est_filepath` by `CBlockPolicyEstimator::m_estimation_filepath` decouples this class from `ArgsManager gArgs`.

ariard reviewed (approved):
Code Review ACK 1670767

Changes since last review, adding more bytes suffix, switching `m_expiry` from std::chrono::hours to std::chrono::seconds and correcting some comments from kb to vkb.

ryanofsky commented on src/mempool_limits.h:
```diff
@@ -0,0 +1,23 @@
+ *
+ * Most of the time, this struct should be referenced as CTxMemPool::Limits.
+ */
+struct MemPoolLimits {
```
In commit "mempool: Introduce (still-unused) MemPoolLimits" (6843403a15cab34bc4e1e7acf7a0e509a2d437df)

Might be good to document these members using documentation that is deleted in later commit "mempool: Use m_limit for UpdateTransactionsFromBlock" (7375ec6b6ada8f45b585f272ab20f1b9c1328ed6). Like for "The maximum allowed number of transactions including the entry and its ancestors" interesting to know the count is inclusive.

ryanofsky reviewed (commented):
Conditional code review ACK 167076739f02ddcf754200b378d504d320d0ff22 if negative mempool size InitError is fixed: https://github.com/bitcoin/bitcoin/pull/25290#discussion_r903619629. 

I'd probably fix this by just making option types `int64_t` instead of `size_t` since GetintArg is returning `int64_t` anyway. This would also facilitate moving error handling into the mempool constructor in the future (https://github.com/bitcoin/bitcoin/pull/25290#discussion_r900272797, https://github.com/bitcoin/bitcoin/pull/25290#discussion_r900420112)

dongcarl commented on src/util/system.cpp:
```diff
@@ -612,36 +612,86 @@ bool ArgsManager::IsArgNegated(const std::string& strArg) const
+        return value.getValStr();
+    } else {
+        return value.get_str();
+    }
```
Fixed as of fc02f77ca6

dongcarl commented on src/util/system.cpp:
```diff
@@ -612,36 +612,86 @@ bool ArgsManager::IsArgNegated(const std::string& strArg) const
+        return value.getInt<int64_t>();
+    } else {
+        return LocaleIndependentAtoi<int64_t>(value.get_str());
+    }
```
Fixed as of fc02f77ca6

dongcarl commented on src/util/system.cpp:
```diff
@@ -612,36 +612,86 @@ bool ArgsManager::IsArgNegated(const std::string& strArg) const
+
+std::optional<bool> SettingToBool(const util::SettingsValue& value)
+{
+    return value.isNull() ? std::nullopt : std::make_optional(value.isBool() ? value.get_bool() : InterpretBool(value.get_str()));
```
Fixed as of fc02f77ca6

dongcarl commented on src/mempool_args.cpp:
```diff
@@ -10,4 +10,6 @@
 {
     mempool_opts.check_ratio = argsman.GetIntArg("-checkmempool", mempool_opts.check_ratio);
+
+    if (auto mb = argsman.GetIntArg("-maxmempool")) mempool_opts.max_size_bytes = *mb * 1000000;
```
Fixed as of 468ee5b9a8

dongcarl commented on src/mempool_options.h:
```diff
@@ -18,6 +23,7 @@ struct MemPoolOptions {
     CBlockPolicyEstimator* const estimator = nullptr;
     /* The ratio used to determine how often sanity checks will run.  */
     int check_ratio = 0;
+    size_t max_size_bytes = DEFAULT_MAX_MEMPOOL_SIZE_MB * 1000000;
```
Fixed as of 468ee5b9a8

dongcarl commented on src/txmempool.h:
```diff
@@ -701,6 +703,9 @@ class CTxMemPool
+    CFeeRate GetMinFee() const {
+        return GetMinFee(m_max_size_bytes);
+    }
     CFeeRate GetMinFee(size_t sizelimit) const;
```
It seems that only the version of `GetMinFee` with a custom size limit is test-only... Perhaps we make that version protected in `CTxMemPool` and have the tests use a:

```cpp
class CMemPoolTest : public CTxMemPool
{
public:
    using CTxMemPool::GetMinFee;
};
```

Or did you have something else in mind? I'm hoping it's just a small change here since it's already existing behaviour...

MarcoFalke replied:
`class MemPoolTest` sounds fine

maflcko reviewed (approved):
Some style fixups, but the bug is still there, I think.


re-ACK 468ee5b9a8 🎨

Changes:
* Style fixups (less verbose code) in the first commit
* Set mempool_opts.estimator = nullptr in the tx_pool fuzz target
* Style: Use 1'000'000 over 1000000, and 1'000 over 1000
* Style: Less verbose code in getPackageLimits
* Doc: Remove leftover code comment

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 468ee5b9a8 🎨

Changes:
* Style fixups (less verbose code) in the first commit
* Set mempool_opts.estimator = nullptr in the tx_pool fuzz target
* Style: Use 1'000'000 over 1000000, and 1'000 over 1000
* Style: Less verbose code in getPackageLimits
* Doc: Remove leftover code comment
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjJswwAyoo1eM+cuPf2dyA2ZXSC9fYwq0geBB/rpaUzr1tYTPiD1NTDcKcYEPh5
MkFgcPzCl65QbxbLsqH2H6WQ5y8ITuYsL6tqQaqqgg+4XLAK1fIO0mBH4/nGmivv
8Tq5AMolDwqroSYhHqa4x4sVnOg/YG6mCi+BzqgfR4hRRWbGayTTKAapm6LCyLmI
anMlIh6jItYQ7Vzy5HUPkQrNsxad29FrG8Jldy9Sfz8kbz84KRe3D/yHsL5bzkYH
TOsPBjouLQK7YEXQiqri/3SSCvO557zhI1Fg3CiEuoZrYLz6+PxNTba40kVxV2X/
3oVQAObfQ5a2D60sVYKeaMweLs2xx/+9lFojD7AyYEeobEqMi+nCJipAdvv3EEHU
1doqX9XyMw/E1PQXCb73l2YGpp5Q1ear66584cIMYT4vM2UXy0jF+a4lpFBSDCT+
nk2tNSiH6JcZUj+LT8i9tCZoQBonf5thKUBphoh9w0E5TFYdDboiDyPGM9Q8DSN7
sA1kGAEa
=32vq
-----END PGP SIGNATURE-----
```


</details>

dongcarl committed:
fuzz/rbf: Add missing TestingSetup

MarcoFalke mentioned that this is likely a bug since "any log messages
should be muted, not accumulated and turned into an OOM when fuzzing for
a long time".

dongcarl commented:
I'm having some challenges with using `int64_t` vs `size_t` for `CTxMemPool::Options::max_size_bytes`. From my cursory reading, it seems like overflow of signed integral types are UB, so I'm not sure what the first check here is doing:

https://github.com/bitcoin/bitcoin/blob/1da1c0dd66265b6a610bf6c21bf2090910cb4fb3/src/init.cpp#L934

Is it to somehow catch the platforms where the apparent behaviour is to wrap around to negative when signed integral types overflow? That doesn't seem right to me.

In my mind, ideally we would eventually have `ApplyArgsManOptions` check that the command line arguments that require multiplication fit within their destination integral types after multiplication. This is a safety check that's not done even prior to this PR.

There's also the fact that prior to this PR, `-maxmempool` was passed to many functions as `size_t`, which could have underflowed...

Given that this PR is already a little bloaty, how about this: we use `int64_t` wherever we replace `GetIntArg` for now, which will make sure that there's no unintentional behaviour change, and we follow up with a PR adding bounds checked multiplication in `ApplyArgsManOptions` and propagate errors up?

MarcoFalke commented on src/test/fuzz/rbf.cpp:
```diff
@@ -23,7 +24,7 @@ FUZZ_TARGET(rbf)
         return;
     }
-    CTxMemPool pool;
+    CTxMemPool pool{{}};
```
Closing as resolved

jamesob commented on src/validation.cpp:
```diff
@@ -2298,7 +2296,7 @@ CoinsCacheSizeState CChainState::GetCoinsCacheSizeState()
     return this->GetCoinsCacheSizeState(
         m_coinstip_cache_size_bytes,
-        gArgs.GetIntArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE_MB) * 1000000);
+        m_mempool ? m_mempool->m_max_size : 0);
```
I think as-written, this is the right behavior. The argument passed here (see https://github.com/bitcoin/bitcoin/pull/16945/commits/b17e91d842724d2888a179a73585cc4c2ef1dc21#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R2203-R2204 for more context) ultimately just accounts for the fact that we repurpose "extra" mempool headroom for the coinscache (see https://github.com/bitcoin/bitcoin/blob/master/src/init.cpp#L1425).

dongcarl replied:
Thanks for the help James! 😄

ryanofsky reviewed (approved):
Code review ACK 7e4f61b2901c8e1506a7b3246a4518e2aa4061f5. Many cleanups since last review, main one is constructing mempools more consistently in test. Also `max_size_bytes` and `MemPoolLimits` are now `int64_t` instead of `size_t` to avoid an unintended change in behavior. `m_max_size_bytes` is still `size_t`

---

re: https://github.com/bitcoin/bitcoin/pull/25290#issuecomment-1165951176

> it seems like overflow of signed integral types are UB, so I'm not sure what the first check here is doing:
> https://github.com/bitcoin/bitcoin/blob/1da1c0dd66265b6a610bf6c21bf2090910cb4fb3/src/init.cpp#L932-L934

I think it's just needed to show an error in the case where both options are passed negative values.

dongcarl commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -121,6 +124,19 @@ void MockTime(FuzzedDataProvider& fuzzed_data_provider, const CChainState& chain
+
+    // ...override specific options for this specific fuzz suite
+    mempool_opts.estimator = nullptr;
+    mempool_opts.check_ratio = 1;
```
For `.estimator`, it's overriding since `MemPoolOptionsForTest` would have set it.
For `.check_ratio`, yeah it's setting the same value, but it's essentially being explicit in case some day `MemPoolOptionsForTest` is changed.

dongcarl commented:
@MarcoFalke 

> in [b61080b](https://github.com/bitcoin/bitcoin/commit/b61080b1993c10d953352701a54067955ca7401b): Please add a _**not**_, see my original comment, which does have a "not": [#25290 (comment)](https://github.com/bitcoin/bitcoin/pull/25290#discussion_r903529958)

😅 That is definitely my bad, fixed in latest push!

dongcarl commented on src/init.cpp:
```diff
@@ -1434,6 +1430,9 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     };
     ApplyArgsManOptions(args, mempool_opts);
     mempool_opts.check_ratio = std::clamp<int>(mempool_opts.check_ratio, 0, 1000000);
+    if (mempool_opts.max_size < 0 || mempool_opts.max_size < mempool_opts.limits.descendant_size) {
```
Fixed by making `mempool_opts.max_size_bytes` a `int64_t` to avoid an unintended change in behavior

dongcarl commented on src/test/util/setup_common.cpp:
```diff
@@ -161,7 +164,12 @@ ChainTestingSetup::ChainTestingSetup(const std::string& chainName, const std::ve
-    m_node.mempool = std::make_unique<CTxMemPool>(m_node.fee_estimator.get(), m_node.args->GetIntArg("-checkmempool", 1));
+    CTxMemPool::Options mempool_opts{
+        Desig(estimator) m_node.fee_estimator.get(),
+        Desig(check_ratio) 1,
```
Fixed as of c7dbe3ae368096f1b3111e0ef7e08a7a5b4ee2a7

dongcarl commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -142,7 +152,7 @@ FUZZ_TARGET_INIT(tx_pool_standard, initialize_tx_pool)
     constexpr CAmount SUPPLY_TOTAL{COINBASE_MATURITY * 50 * COIN};
 
-    CTxMemPool tx_pool_{/*estimator=*/nullptr, /*check_ratio=*/1};
+    CTxMemPool tx_pool_{MakeMempool(*node.args)};
```
Closing, feel free to reopen if still a problem!

dongcarl commented on src/test/fuzz/tx_pool.cpp:
```diff
@@ -121,6 +123,14 @@ void MockTime(FuzzedDataProvider& fuzzed_data_provider, const CChainState& chain
+static inline CTxMemPool MakeMempool(const ArgsManager& argsman)
+{
+    CTxMemPool::Options mempool_opts{};
+    ApplyArgsManOptions(argsman, mempool_opts);
```
Fixed as of c7dbe3ae368096f1b3111e0ef7e08a7a5b4ee2a7

dongcarl commented on src/test/fuzz/policy_estimator_io.cpp:
```diff
@@ -22,7 +22,7 @@ FUZZ_TARGET_INIT(policy_estimator_io, initialize_policy_estimator_io)
     CAutoFile fuzzed_auto_file = fuzzed_auto_file_provider.open();
     // Re-using block_policy_estimator across runs to avoid costly creation of CBlockPolicyEstimator object.
-    static CBlockPolicyEstimator block_policy_estimator;
+    static CBlockPolicyEstimator block_policy_estimator{gArgs.GetDataDirNet() / FEE_ESTIMATES_FILENAME};
```
Fixed as of 4846995a06ee1886693f096f17cce6005b3cf796

dongcarl commented on src/init.cpp:
```diff
@@ -1434,6 +1430,9 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     ApplyArgsManOptions(args, mempool_opts);
     mempool_opts.check_ratio = std::clamp<int>(mempool_opts.check_ratio, 0, 1000000);
+    if (mempool_opts.max_size < 0 || mempool_opts.max_size < mempool_opts.limits.descendant_size) {
+        return InitError(strprintf(_("-maxmempool must be at least %d MB"), std::ceil(mempool_opts.limits.descendant_size / 1000000.0)));
```
See: https://github.com/bitcoin/bitcoin/pull/25290#discussion_r907712185

dongcarl committed:
mempool: Make GetMinFee() with custom size protected

The version of GetMinFee() with a custom size specification is and
should only be used by tests. Mark it as protected and use a derived
class exposing GetMinFee() as public in tests.

dongcarl commented:
Pushed 7e4f61b2901c8e1506a7b3246a4518e2aa4061f5 -> 54913736701ed336a2be94feb0f77fae21226ba4
- Fixed quoting Marco: https://github.com/bitcoin/bitcoin/pull/25290#issuecomment-1167172152
- Added comment about `CChainState::GetCoinsCacheSizeState()` correctness (thanks James!): https://github.com/bitcoin/bitcoin/pull/25290#discussion_r907438710
- Made `CTxMemPool::m_max_size_bytes` also `int64_t` to avoid an unintended change in behavior. Added to project TODO a task to add bounds checking in `ApplyArgsManOptions(..., MemPoolOptions&)` so that we can safely transition `CTxMemPool::{Options,Limits}` to using unsigned types for sizes.

dongcarl commented:
Maintainers: All reviews have been addressed, I think this PR is in a good place for merge pending CI pass.

dongcarl commented on src/rpc/mempool.cpp:
```diff
@@ -657,9 +657,8 @@ UniValue MempoolInfoToJSON(const CTxMemPool& pool)
-    int64_t maxmempool{gArgs.GetIntArg("-maxmempool", DEFAULT_MAX_MEMPOOL_SIZE_MB) * 1000000};
-    ret.pushKV("maxmempool", maxmempool);
-    ret.pushKV("mempoolminfee", ValueFromAmount(std::max(pool.GetMinFee(maxmempool), ::minRelayTxFee).GetFeePerK()));
+    ret.pushKV("maxmempool", (int64_t)pool.m_max_size_bytes);
```
No longer needed as of 5491373670

dongcarl commented:
I believe the CI failure in wallet_reorgsrestore.py is spurious: https://cirrus-ci.com/task/6073861945950208?logs=ci#L560

I've run `FILE_ENV=./ci/test/00_setup_env_native_asan.sh ./ci/test_run_all.sh` locally and am unable to reproduce the error.

ariard reviewed (approved):
Code Review ACK 5491373

Changes since last review:
- moving `mempool_optjons.h` under the `kernel` namespace
- adding `MemPoolOptionsForTest` in `src/test/util/setup_common.cpp`
- initializing `CTxMemPool` with `MemPoolOptionsForTest`
- moving `MemPoolOptions::max_size_bytes` from `size_t` to `int64_t`
- moving `CTxMemPool::m_max_size_bytes` from `size_t` to `int64_t`
- renaming `DEFAULT_MEMPOOL_EXPIRY` to `DEFAULT_MEMPOOL_EXPIRY_HOURS`
- using ternary in `getPackageLimits()`
- adding comments in `MemPoolLimits`
- making `descendant_limits_bytes` in `src/init.cpp` a `int64_t`
- adding new `policy/fes_args.cpp` and the fee estimator path constructor `FeeestPath`

glozow reviewed (commented):
light code review ACK 5491373670. Very happy with the reduced querying of gArgs for mempool options, agree with the code organization, I'm much less familiar with init than mempool but stepped through each commit and the code looks correct to me.

laanwj commented on src/util/system.h:
```diff
@@ -365,6 +372,7 @@ class ArgsManager
      * @return command-line argument or default value
      */
     bool GetBoolArg(const std::string& strArg, bool fDefault) const;
+    std::optional<bool> GetBoolArg(const std::string& strArg) const;
```
I like the idea of this commit.
My original idea for the future of default arguments was to make ArgsManager handle them internally at registration time, and get rid of the `default` arguments to these functions in that way. But this works.

laanwj commented on src/policy/fees_args.h:
```diff
@@ -0,0 +1,15 @@
+
+fs::path FeeestPath(const ArgsManager& argsman);
+
+
```
nit: extra newline?

dongcarl replied:
Fixed as of b96e009a8a36f4feeee49d1a3c8f85fb9ab0d920

laanwj commented on src/policy/fees_args.h:
```diff
@@ -0,0 +1,15 @@
+
+class ArgsManager;
+
+fs::path FeeestPath(const ArgsManager& argsman);
```
Please add a doc comment for this function. E.g.
```
/** Return path of fee estimates data file. */
```

dongcarl replied:
Fixed as of b96e009a8a36f4feeee49d1a3c8f85fb9ab0d920

laanwj commented on src/test/fuzz/policy_estimator_io.cpp:
```diff
@@ -11,9 +12,14 @@
 void initialize_policy_estimator_io()
 {
     static const auto testing_setup = MakeNoLogFileContext<>();
+    g_setup = testing_setup.get();
```
Having a global dangling pointer here is kind of ugly. Is there really no other way to get this information in?
If you need a global, what about storing the path in the global instead of a pointer. It seems slightly safer at least.

MarcoFalke replied:
Yeah, this is the general pattern. Let me know if there is a better way.

And there should be no risk in doing "risky" stuff in the unit or fuzz tests, such as globals, raw pointers/memory, "wild" casts, etc..., as they are run in sanitizers that should detect any issue if it arises. (As opposed to real code, which is not 100% covered by sanitizers)

laanwj replied:
Ok, that's a good point. I was mainly afraid of false positives/negatives due to pointers pointing to the wrong memory. But yes, the sanitizers should catch that.

MarcoFalke commented on src/validation.cpp:
```diff
@@ -4645,7 +4641,7 @@ static const uint64_t MEMPOOL_DUMP_VERSION = 1;
 bool LoadMempool(CTxMemPool& pool, CChainState& active_chainstate, FopenFn mockable_fopen_function)
 {
-    int64_t nExpiryTimeout = gArgs.GetIntArg("-mempoolexpiry", DEFAULT_MEMPOOL_EXPIRY) * 60 * 60;
+    int64_t nExpiryTimeout = std::chrono::seconds{pool.m_expiry}.count();
```
nit in e741c49716a65b624fcf91db0de6dd11617675ad: Might be nice to make `nExpiryTimeout` type-safe in a follow-up.

dongcarl replied:
True! Not addressing in this PR.

MarcoFalke commented on src/kernel/mempool_limits.h:
```diff
@@ -0,0 +1,25 @@
+#define BITCOIN_KERNEL_MEMPOOL_LIMITS_H
+
+#include <cstdint>
+#include <policy/policy.h>
```
nit in 1ad4268e1a7da0e97a97fa677f7b50f226d084b4: Usually std-includes come last in a separate paragraph

dongcarl replied:
Fixed in 9333427014

MarcoFalke commented on src/txmempool.cpp:
```diff
@@ -148,15 +147,15 @@ void CTxMemPool::UpdateForDescendants(txiter updateIt, cacheMap& cachedDescendan
 }
 
-void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashesToUpdate, uint64_t ancestor_size_limit, uint64_t ancestor_count_limit)
+void CTxMemPool::UpdateTransactionsFromBlock(const std::vector<uint256> &vHashesToUpdate)
```
nit in 5a3d18eb2c12e046d35d628dc67daa9e2bb343b3: Can update format while touching this line?

dongcarl replied:
Fixed in 6c5c60c412

MarcoFalke commented on src/init.cpp:
```diff
@@ -1433,6 +1428,11 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     ApplyArgsManOptions(args, mempool_opts);
     mempool_opts.check_ratio = std::clamp<int>(mempool_opts.check_ratio, 0, 1000000);
+
+    int64_t descendant_limit_bytes = mempool_opts.limits.descendant_size_vbytes * 40;
```
nit in 470347d2ffd462d635fdf344e38332a4eb4b8ad3: could add a `{}` scope to avoid leaking this symbol into the full `AppInitMain` scope?

dongcarl replied:
I'm not going to do this in this PR, feel free to follow up.

maflcko reviewed (approved):
lgtm

review ACK 54913736701ed336a2be94feb0f77fae21226ba4 📏

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 54913736701ed336a2be94feb0f77fae21226ba4 📏
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUijogv8D4DYVC5hv7qOlIEMk+Yj7ZthU8aCSgXn6bhN611QGGuj40G9dfTNtOZ5
oLKTttz9W1ML68IdlN+syYL112EN9lcJi+jls5wdof3dr1qrKZJEdBEi/sL+CBjt
h1vr8zHHWSCw2IYhEtHmQMj8oNbtID4NCG3ngyAEh9xzw7UGrzM9bxUI0qJhQJ1n
/AXxVZZ32sM+5G2YYa1Pc+HoxmpcPhZuF5sWCxTDGeDkwA8DEbvyJUOwP+ibO65x
hMDvsoLIwYsVkXCw9LQFG2nUB2uWKujmi4AgGqsOr/+mO6foxsMqwrZLhXyZn39U
JWBtkWkizU69fYMBdr5+ku0BdH6c6EL/esfsTQ8n16W+Dlq/1Pm0sgVwQmGOF8F2
rdrxKl4YSodH7vX8ChNxNxaASBqKoLltSOxJ7QWdB9py4F7c1yd/dAA1GpUVz8g6
BVcxDbpa0JTkYU2a3yUx02vs3SUPb+P2ZTz+gsxbDmw/gT/J8RZ3RDwNQKafhTzZ
9imCuWxg
=vOeM
-----END PGP SIGNATURE-----
```


</details>

ryanofsky reviewed (approved):
Code review ACK b96e009a8a36f4feeee49d1a3c8f85fb9ab0d920. Commit message and comment cleanups since last review, and m_max_size_bytes is changed from size_t to int64_t

dongcarl commented:
Pushed b96e009a8a36f4feeee49d1a3c8f85fb9ab0d920 -> d1684beabe5b738c2cc83de83e1aaef11a761b69
- Addressed some of Marco's review comments

dongcarl commented:
I'm likely not going to repush for nits from this point forwards, since there's quite a lot of Code Review ACKs already and I want to reduce re-review burden. I hope that's understandable. We can always follow up in future PRs.

dongcarl commented on src/txmempool.cpp:
```diff
@@ -451,8 +451,9 @@ void CTxMemPoolEntry::UpdateAncestorState(int64_t modifySize, CAmount modifyFee,
-    : m_check_ratio(check_ratio), minerPolicyEstimator(estimator)
+CTxMemPool::CTxMemPool(const Options& opts)
+    : m_check_ratio(opts.check_ratio),
+      minerPolicyEstimator(opts.estimator)
```
Fixed in f1941e8bfd and others

dongcarl commented on src/init.cpp:
```diff
@@ -1426,10 +1428,16 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        Desig(check_ratio) chainparams.DefaultConsistencyChecks() ? 1 : 0,
+    };
+    ApplyArgsManOptions(args, mempool_opts);
+    mempool_opts.check_ratio = std::clamp<int>(mempool_opts.check_ratio, 0, 1000000);
```
Fixed in f1941e8bfd

dongcarl commented on src/test/mempool_tests.cpp:
```diff
@@ -16,6 +16,12 @@ BOOST_FIXTURE_TEST_SUITE(mempool_tests, TestingSetup)
 
```
Fixed in 386c9472c8

ryanofsky reviewed (approved):
Code review ACK d1684beabe5b738c2cc83de83e1aaef11a761b69. Just minor cleanups since last review, mostly switching to brace initialization

MarcoFalke commented:
ACK d1684beabe5b738c2cc83de83e1aaef11a761b69 🍜

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK d1684beabe5b738c2cc83de83e1aaef11a761b69 🍜
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYEACgkQzit1aX5p
pUgupwv9G/2wfNggsXuzCwHm6yNWsM5mmsQMgwhxAAWWFG0O+Wa3rbImNzW04Qp5
LcHOWPiZ4nx/aLL5TtvzEPYJ5iy3inLU/ekMLEzRoRJgdWaPAnXUzXEyK6/wXAie
J2Oa1RZTpJ/HuaqnrQ8grupoFHkLw7RQ3j4uhOn1aROUREz+C4JragufIQtXCnMJ
EBAMn0MnNo3TWmeLEVxdpjkk6Hhuln4Kqv1/MGo9g+CLu6zFsLssBMCO2lGP/2up
5p/HKyPVYh/8lqBjGW2JH7YsHj5pnS614okonpzOEKN6Qktx/YAXnnMjJyW3SvmP
lVdYjNW38nekegQNSiceliYQauO6SiwmmEL4r921ytAR/zKjYEvKrpj2+w8XclU2
s/KGZX2x9LqtxkL7rK0HBTqOL4ELmcp2cDqeYiPnL5FHgmERODnE+uACNszpYCHR
waRWtqkJE5tVYn7gJ3frICTDUWQFbcR5+ofdDDQPWahKfqZwSiRAd68R7x4tL6Dp
PKfpIgo5
=seCw
-----END PGP SIGNATURE-----
```


</details>

MarcoFalke commented:
Let me know if you want to follow-up for https://github.com/bitcoin/bitcoin/pull/25290#discussion_r908811064 or want me to do it. I see that you have a follow-up already open, so I don't want to duplicate work or queue up merge conflicts.

dongcarl commented:
@MarcoFalke Answered in: https://github.com/bitcoin/bitcoin/pull/25487#issuecomment-1170132932
